optimum/habana/transformers/models/gptj/modeling_gptj.py

```diff
@@ -359,6 +354,19 @@ def gaudi_gptj_model_forward(
     presents = () if use_cache else None
     all_self_attentions = () if output_attentions else None
     all_hidden_states = () if output_hidden_states else None
+    rotary_dim = self.config.rotary_dim
+    embed_dim = self.config.hidden_size
+    pos_embd_dim = rotary_dim or embed_dim
+    max_positions = self.config.max_position_embeddings
+    embed_positions = create_sinusoidal_positions(max_positions, pos_embd_dim).to(torch.bfloat16)
+    embed_positions = embed_positions.repeat(position_ids.shape[0], 1, 1)
+    if embed_positions.device != position_ids.device:
+        embed_positions = embed_positions.to(position_ids.device)
+    repeated_position_ids = position_ids.unsqueeze(-1).repeat(1, 1, embed_positions.shape[-1])
+    sincos = torch.gather(embed_positions, 1, repeated_position_ids)
+    sin, cos = torch.split(sincos, sincos.shape[-1] // 2, dim=-1)
+    sin = sin.contiguous()
+    cos = cos.contiguous()
```

Comment:
I understand that this piece of code comes from the code blocks that were removed above. Could this be moved to a dedicated method that would be called here please?

Reply:
Theoretically it can be done, but test shows that an additional memcpy occurred, perf drop detail as follow.
Throughput (including tokenization) = 3885.6094019038055 tokens/second
Memory allocated                    = 27.33 GB
Max memory allocated                = 28.73 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 8.958231755999805 seconds

Reply:
> I understand that this piece of code comes from the code blocks that were removed above. Could this be moved to a dedicated method that would be called here please?
FYI, changes looks like this
`def get_embed_positions(embed_positions, position_ids):
    embed_positions = embed_positions.repeat(position_ids.shape[0], 1, 1)
    if embed_positions.device != position_ids.device:
        embed_positions = embed_positions.to(position_ids.device)
    return embed_positions`

Reply:
That's surprising as objects are passed to functions by references if I'm not mistaken.
Okay, in that case, could you just add a comment above this block saying which methods it replaces, and also add a blank line right above and below please?

Reply:
> That's surprising as objects are passed to functions by references if I'm not mistaken. Okay, in that case, could you just add a comment above this block saying which methods it replaces, and also add a blank line right above and below please?

Surprising indeed. Anyway, changed accordingly.

Reply:
> That's surprising as objects are passed to functions by references if I'm not mistaken. Okay, in that case, could you just add a comment above this block saying which methods it replaces, and also add a blank line right above and below please?

By the way, I think the `make style` removed blank line I added below this block : )